profile-widget/profilewidget2.cpp

```diff
@@ -2044,70 +2045,75 @@ void ProfileWidget2::clearPictures()
+		it->thumbnail->setPixmap(QPixmap::fromImage(thumbnail.scaled(size, size, Qt::KeepAspectRatio)));
 	}
 }
 
+bool ProfileWidget2::PictureEntry::operator< (const PictureEntry &e) const
+{
+	return std::tie(offset.seconds, filename) < std::tie(e.offset.seconds, e.filename);
+}
+
 void ProfileWidget2::plotPictures()
 {
-	DivePictureModel *m = DivePictureModel::instance();
-	pictures.resize(m->rowDDEnd - m->rowDDStart);
+	pictures.clear();
+	if (currentState == ADD || currentState == PLAN)
+		return;
+	FOR_EACH_PICTURE(current_dive)
+		pictures.push_back(PictureEntry{ picture->offset, QString(picture->filename),  {} });
+	if (pictures.empty())
+		return;
+	std::sort(pictures.begin(), pictures.end());
 
 	double x, y, lastX = -1.0, lastY = -1.0;
-	for (int i = m->rowDDStart; i < m->rowDDEnd; i++) {
-		int picItemNr = i - m->rowDDStart;
-		int offsetSeconds = m->index(i, 1).data(Qt::UserRole).value<int>();
-		// it's a correct picture, but doesn't have a timestamp: only show on the widget near the
-		// information area. A null pointer in the pictures array indicates that this picture is not
-		// shown.
-		if (!offsetSeconds) {
-			pictures[picItemNr].reset();
+	int size = Thumbnailer::defaultThumbnailSize();
+	for (PictureEntry &e: pictures) {
+		// Only show pictures whose offsets are withing the profile
+		if (e.offset.seconds <= 0 || e.offset.seconds >= current_dive->duration.seconds)
 			continue;
-		}
-		DivePictureItem *item = pictures[picItemNr].get();
-		if (!item) {
-			item = new DivePictureItem;
-			pictures[picItemNr].reset(item);
-			scene()->addItem(item);
-		}
-		item->setVisible(prefs.show_pictures_in_profile);
-		item->setPixmap(m->index(i, 0).data(Qt::UserRole).value<QPixmap>());
-		item->setFileUrl(m->index(i, 1).data().toString());
+		e.thumbnail.reset(new DivePictureItem);
+		scene()->addItem(e.thumbnail.get());
+		e.thumbnail->setVisible(prefs.show_pictures_in_profile);
+		QImage thumbnail = Thumbnailer::instance()->fetchThumbnail(e.filename).scaled(size, size, Qt::KeepAspectRatio);
+		e.thumbnail->setPixmap(QPixmap::fromImage(thumbnail));
+		e.thumbnail->setFileUrl(e.filename);
 		// let's put the picture at the correct time, but at a fixed "depth" on the profile
 		// not sure this is ideal, but it seems to look right.
-		x = timeAxis->posAtValue(offsetSeconds);
-		if (i == 0)
+		x = timeAxis->posAtValue(e.offset.seconds);
+		if (lastX < 0.0)
 			y = 10;
 		else if (fabs(x - lastX) < 3 && lastY <= (10 + 14 * 3))
 			y = lastY + 3;
 		else
 			y = 10;
 		lastX = x;
 		lastY = y;
-		item->setPos(x, y);
+		e.thumbnail->setPos(x, y);
 	}
 }
 
-void ProfileWidget2::removePictures(const QModelIndex &, int first, int last)
+// TODO: This does not check for the fact that the same image may be attributed
+// to different dives! Deleting the picture from one dive may therefore remove
+// it from the profile of a different dive.
+void ProfileWidget2::removePictures(const QVector<QString> &fileUrls)
```

Comment:
i think this TODO has to be fixed before we marge, so that we do not forget about it and a user sends us a bug report.